src/scripting/mapgen_lua_kernel.cpp

```diff
@@ -295,16 +295,40 @@ mapgen_lua_kernel::mapgen_lua_kernel(const config* vars)
 	cmd_log_ << lua_terrainfilter::register_metatables(L);
 }
 
-void mapgen_lua_kernel::run_generator(const char * prog, const config & generator)
+bool mapgen_lua_kernel::run_generator(std::string_view file, const char * field, const config & generator)
 {
-	load_string(prog, "", std::bind(&lua_kernel_base::throw_exception, this, std::placeholders::_1, std::placeholders::_2));
+	luaW_getglobal(mState, "wesnoth", "require");
+	lua_push(mState, file);
+	if(!protected_call(1, 1)) {
+		cmd_log_ << "Error: Failed to load '" << file << "'mapgen.\n";
+	}
+	lua_getfield(mState, -1, field);
+	if(lua_isnoneornil(mState, -1)) {
+		lua_pop(mState, 1);
+		return false;
+	}
 	luaW_pushconfig(mState, generator);
 	protected_call(1, 1, std::bind(&lua_kernel_base::throw_exception, this, std::placeholders::_1, std::placeholders::_2));
+	return true;
```

Comment:
This seems to indicate an issue, though I'm not sure if the issue is actually here.

From what I can tell, what this ultimately means is that even if the function raises any errors, `run_generator` considers it to be a success. (I suppose, it was successfully called even if it didn't successfully complete.)

But in the case of the `lua` generator, that's not entirely correct. I'm not sure whether this line is wrong or if it's the specific generator that needs to change, but if a scenario with `map_generator=lua` is used to generate a map, and the config contains `create_scenario` but not `create_map`, currently it looks like the fallback scenario will break – it will call `create_map` from `lua.lua`, which will try to load an empty string, which will fail and raise an error. And then this will return true and it won't fall back to `create_scenario` as it's supposed to.

Reply:
> From what I can tell, what this ultimately means is that even if the function raises any errors, run_generator considers it to be a success. (I suppose, it was successfully called even if it didn't successfully complete.)

I'd think that passing `lua_kernel_base::throw_exception`  mean that on eeor well get an exception.


> but if a scenario with map_generator=lua is used to generate a map, and the config contains create_scenario but not create_map, currently it looks like the fallback scenario will break – it will call create_map from lua.lua, which will try to load an empty string

It it doesn't contain create_map then the check above this for create_map being noneOrNil will fail and this function will return false.

Reply:
Yeah, it's not that the generator doesn't contain `create_map` but that the _config_ doesn't. I wasn't sure exactly, but I did think it might be an issue in `lua.lua` rather than here.